coresdk/src/coresdk/database.h

```diff
@@ -1,60 +1,301 @@
 
+/**
+ * @brief Loads or creates, and returns a database
+ *
+ * The supplied `filename` is the name of the file of the database. The
+ * supplied `name` indicates the name to use to refer to this `database`.
+ * The `database` can then be retrieved by passing this `name` to
+ * the `database_named` function.
+ *
+ * @param name      The name used to refer to the database.
+ * @param filename  The filename used to locate the database to use.
+ *
+ * @returns A new `database` with the initialised values provided.
+ *
+ * @attribute class       database
+ * @attribute constructor true
+ */
 database open_database(string name, string filename);
 
+/**
+ * Returns the `query_result` for the query passed into `sql`.
+ *
+ * @param db The database to perform `sql` on.
+ *
+ * @param sql The sql statement to perform on `db`.
+ *
+ * @returns Returns the `query_result` which represents
+ * the result of perfoming `sql` on `db`.
+ *
+ * @attribute class   database
+ * @attribute method  run_sql_on_db
+ */
 query_result run_sql(database db, string sql);
 
+/**
+ * Returns the `query_result` for the query passed into `sql`.
+ *
+ * @param database_name The string name of the database to perform `sql` on.
+ *
+ * @param sql The sql statement to perform on the database at `database_name`.
+ *
+ * @returns Returns the `query_result` which represents
+ * the result of perfoming `sql` on the database at `database_name`.
+ *
+ * @attribute class   database
+ * @attribute method  run_sql_on_database_named
+ */
 query_result run_sql(string database_name, string sql);
 
+/**
+ * Frees all of the databases which have been loaded.
+ *
+ * @attribute class   database
+ * @attribute method  free_all_query_results
+ */
 void free_all_query_results();
 
+/**
+ * Releases the SplashKit resources associated with the query.
+ *
+ * @param query The `query_result` whose resources should be released.
+ *
+ * @attribute class       database
+ * @attribute self        query_result
+ */
 void free_query_result(query_result query);
 
+/**
+ * Gets the next row available on a given `query_result`.
+ *
+ * @param result The `query_result` to move to the next row on.
+ *
+ * @returns Returns a `bool` which represents if there was a valid row of data to move to.
+ *
+ * @attribute class   database
+ * @attribute method  get_next_row
+ */
 bool get_next_row(query_result result);
 
+/**
+ * Checks if a given `query_result` has a valid row of data.
+ *
+ * @param result The `query_result` to check for a valid row of data on.
+ *
+ * @returns Returns a `bool` which represents if there was a valid row of data.
+ *
+ * @attribute class   database
+ * @attribute method  has_row
+ */
 bool has_row(query_result result);
 
+/**
+ * Resets a query_result back to its initial state, ready to be re-executed.
+ *
+ * @param result The `query_result` to reset.
+ *
+ * @attribute class   database
+ * @attribute method  reset_query_result
+ */
 void reset_query_result(query_result result);
```

Comment:
Maybe this should be under the `QueryResult` class, in which case the `class` would be `query_result`. Then the `method` would be `reset` and we have a `self` attribute of `result`:

```cpp
queryResult.reset();
```

Or would this be introducing too many classes... what do you think @macite?